Source/Private/ASTextNode2.mm

```diff
@@ -0,0 +1,1122 @@
+  //		if (renderer.currentScaleFactor > 0 && renderer.currentScaleFactor < 1.0) {
+  //			// while not perfect, this is a good estimate of what the ascender of the scaled font will be.
+  //			self.style.ascender *= renderer.currentScaleFactor;
+  //			self.style.descender *= renderer.currentScaleFactor;
+  //		}
+  //	}
+  //
+  //	// Add the constrained size back textContainerInset
+  //	size.width += (_textContainerInset.left + _textContainerInset.right);
+  //	size.height += (_textContainerInset.top + _textContainerInset.bottom);
+  //
+  //	return CGSizeMake(std::fmin(size.width, originalConstrainedSize.width),
+  //					  std::fmin(size.height, originalConstrainedSize.height));
+}
+
+#pragma mark - Modifying User Text
+
+// Returns the ascender of the first character in attributedString by also including the line height if specified in paragraph style.
++ (CGFloat)ascenderWithAttributedString:(NSAttributedString *)attributedString
+{
+  UIFont *font = [attributedString attribute:NSFontAttributeName atIndex:0 effectiveRange:NULL];
+  NSParagraphStyle *paragraphStyle = [attributedString attribute:NSParagraphStyleAttributeName atIndex:0 effectiveRange:NULL];
+  if (!paragraphStyle) {
+    return font.ascender;
+  }
+  CGFloat lineHeight = MAX(font.lineHeight, paragraphStyle.minimumLineHeight);
+  if (paragraphStyle.maximumLineHeight > 0) {
+    lineHeight = MIN(lineHeight, paragraphStyle.maximumLineHeight);
+  }
+  return lineHeight + font.descender;
+}
+
+- (NSAttributedString *)attributedText
+{
+  ASDN::MutexLocker l(__instanceLock__);
+  return _attributedText;
+}
+
+- (void)setAttributedText:(NSAttributedString *)attributedText
+{
+  
+  if (attributedText == nil) {
+    attributedText = [[NSAttributedString alloc] initWithString:@"" attributes:nil];
+  }
+  
+  // Don't hold textLock for too long.
+  {
+    ASDN::MutexLocker l(__instanceLock__);
+    if (ASObjectIsEqual(attributedText, _attributedText)) {
+      return;
+    }
+    
+    _attributedText = attributedText;
+#if AS_TEXTNODE2_RECORD_ATTRIBUTED_STRINGS
+    [ASTextNode _registerAttributedText:_attributedText];
+#endif
+  }
+  
+  // Since truncation text matches style of attributedText, invalidate it now.
+  [self _invalidateTruncationText];
+  
+  NSUInteger length = attributedText.length;
+  if (length > 0) {
+    self.style.ascender = [[self class] ascenderWithAttributedString:attributedText];
+    self.style.descender = [[attributedText attribute:NSFontAttributeName atIndex:attributedText.length - 1 effectiveRange:NULL] descender];
+  }
+  
+  // Tell the display node superclasses that the cached layout is incorrect now
+  [self setNeedsLayout];
+  
+  // Force display to create renderer with new size and redisplay with new string
+  [self setNeedsDisplay];
+  
+  
+  // Accessiblity
+  self.accessibilityLabel = attributedText.string;
+  self.isAccessibilityElement = (length != 0); // We're an accessibility element by default if there is a string.
```

Comment:
Probably not, but is there any way attributedText.length could be different than attributedText.string.length?